### ISSUE TITLE
NC24: InvalidArgumentException: Column "oc_bookmarks"."url" is type String, but exceeding the 4.000 length limit.

### ISSUE THREAD

**Issue by mjz68** [nextcloud/bookmarks#1785]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Meldung des Web-Updaters:

Aktualisierung auf 24.0.0 beta 2

InvalidArgumentException: Column "oc_bookmarks"."url" is type String, but exceeding the 4.000 length limit.
[Detaillierte Protokollmeldungen ]()
Update vorbereiten
Log-Level auf "debug" gesetzt
Wartungsmodus eingeschaltet
Repair MySQL collation
All tables already have the correct collation -> nothing to do
Repair SQLite autoincrement
Copy data from accounts table when migrating from ownCloud
Drop account terms table when migrating from ownCloud
Das Datenbankschema wird aktualisiert
InvalidArgumentException: Column "oc_bookmarks"."url" is type String, but exceeding the 4.000 length limit.

### Steps to reproduce

Web Update

### Expected behavior

Normalen Durchlauf

### Installation method

Manual installation

### Operating system

Debian/Ubuntu

### Web server

Apache (supported)

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### List of activated Apps

```shell
Aktuell nicht möglich
```

**Comment by BombTheDodongos** [comment]
I'm running into the exact same issue upgrading to 24.0.0 beta 2

**Comment by marcelklehr** [comment]
Huh, this breaking change was not mentioned in the v24 issue @kesselb 
Where is this documented?

also see https://github.com/nextcloud/server/issues/31791

**Comment by kesselb** [comment]
> Huh, this breaking change was not mentioned in the v24 issue @kesselb

Indeed. I were not involved in Nextcloud 24 due some health issues and don't know why this new requirement is not documented.

**Comment by nursoda** [comment]
Same here while updating to NC24; system env: Arch, PHP 8.0(!), MariaDB, Redis

**Comment by nursoda** [comment]
Are we sure that this is a bookmarks app issue? To continue with the upgrade 23→24b2, I manually disabled and removed the bookmarks app and re-issued `occ upgrade`. The error remained, so I manually dropped all bookmarks tables. After doing so, the `occ upgrade` now yields

```
InvalidArgumentException: Column "oc_dashboard_data"."data" is type String, but exceeding the 4.000 length limit.
```

Disabling dashboard is possible but does not change the outcome of `occ upgrade`. It is not possible to remove `dashboard` using `occ app:remove dashboard`:

```
dashboard could not be removed as it is a shipped app
```

To continue with the upgrade I manually moved `apps/dashboard` and `apps/dashboardcharts` aside and dumped and dropped `oc_dashboard_data` from the database. Upon `occ upgrade` I now get

```
InvalidArgumentException: Column "oc_external_config"."value" is type String, but exceeding the 4.000 length limit.
```

Disabling `files_external` is possible but does not change the outcome of `occ upgrade`. It is also not possible to remove `files_external` using `occ app:remove files_external` since it is a shipped app – and I probably really should not remove this app. I did anyway (by moving `apps/files_external` aside and dumping/dropping `oc_external_applicable oc_external_config oc_external_mounts oc_external_options`). Next apps the updater complained about were `forms` and `epubreader`. I also uninstalled them and dumped/dropped their tables. After that, I could do the `occ upgrade`.

**Comment by laracroftonline** [comment]
I also have this bug on beta 2

**Comment by Thaurin** [comment]
I also ran into the same issue just now. Is @nursoda's solution a safe workaround for now to continue the upgrade? It seems... risky.

**Comment by nursoda** [comment]
Installing a beta always is. I took the risk to be able to update PHP (to 8.1, I'm on Arch). Turns out that may have not been a good choice: At first it seemed that all instances just ran fine, but one instance is in maintenance mode the second time now for no reason I can think of. Investigating… but most probably these issues are unrelated to cropping the column length. I expect that the migrations would do just that (if the limit is kept).

**Comment by Thaurin** [comment]
Hmm, now I'm considering a complete fresh install, unfortunately. I guess you're right in that it's my own fault for installing a beta, but still. It's a very simple instance and I wonder what went wrong here, and if there's a clean fix. Until then, I can't easily access my files on some devices.

**Comment by Githopp192** [comment]
did  the 24 upgrade on a test vm: 

Current version is 23.0.3.										
										
Update to Nextcloud 24.0.0 beta 2 available. (channel: "beta")										
Following file will be downloaded automatically: https://download.nextcloud.com/server/prereleases/nextcloud-24.0.0beta2.zip										
										
Steps that will be executed:										
[ ] Check for expected files										
[ ] Check for write permissions										
[ ] Create backup										
[ ] Downloading										
[ ] Verify integrity										
[ ] Extracting										
[ ] Enable maintenance mode										
[ ] Replace entry points										
[ ] Delete old files										
[ ] Move new files in place										
[ ] Done										
										
Start update? [y/N] y										
										
Info: Pressing Ctrl-C will finish the currently running step and then stops the updater.										
										
[✔] Check for expected files										
[✔] Check for write permissions										
[✔] Create backup										
[✔] Downloading										
[✔] Verify integrity										
[✔] Extracting										
[✔] Enable maintenance mode										
[✔] Replace entry points										
[✔] Delete old files										
[✔] Move new files in place										
[✔] Done										
										
Update of code successful.										
										
Should the "occ upgrade" command be executed? [Y/n] y										
Nextcloud or one of the apps require upgrade - only a limited number of commands are available										
You may use your browser or the occ upgrade command to do the upgrade										
2022-04-04T22:29:06+02:00 Setting log level to debug										
2022-04-04T22:29:06+02:00 Repair step: Repair MySQL collation										
2022-04-04T22:29:06+02:00 Repair info: All tables already have the correct collation -> nothing to do										
2022-04-04T22:29:06+02:00 Repair step: Repair SQLite autoincrement										
2022-04-04T22:29:06+02:00 Repair step: Copy data from accounts table when migrating from ownCloud										
2022-04-04T22:29:08+02:00 Repair step: Drop account terms table when migrating from ownCloud										
2022-04-04T22:29:08+02:00 Updating database schema										
2022-04-04T22:29:10+02:00 InvalidArgumentException: Column "oc_bookmarks"."url" is type String, but exceeding the 4.000 length limit.										
2022-04-04T22:29:10+02:00 Update failed										
2022-04-04T22:29:10+02:00 Maintenance mode is kept active										
2022-04-04T22:29:10+02:00 Resetting log level										

Operating System: System Linux 4.18.0-305.19.1.el8_4.x86_64 Build => Hat

Webserver: Apache/2.4.37 (centos)

Database: mysql 10.6.4-MariaDB,

PHP version: PHP Version => 7.4.24

[PHP Modules] apcu bcmath bz2 calendar Core ctype curl date dba dom enchant exif fileinfo filter ftp gd gettext gmp hash iconv igbinary imagick imap intl json ldap libxml mbstring mcrypt memcached msgpack mysql mysqli mysqlnd odbc openssl pcntl pcre PDO pdo_mysql PDO_ODBC pdo_pgsql pdo_sqlite pgsql Phar posix readline redis Reflection session shmop SimpleXML soap sockets sodium SPL sqlite3 standard sysvmsg sysvsem sysvshm tokenizer xml xmlreader xmlwriter xsl Zend OPcache zip zlib [Zend Modules] Zend OPcache

Nextcloud: - version: 24.0.0.7

Enabled:
  - admin_audit: 1.13.0
  - bookmarks: 10.2.1
  - circles: 23.1.0
  - cloud_federation_api: 1.6.0
  - comments: 1.13.0
  - contacts: 4.1.0
  - contactsinteraction: 1.4.0
  - cookbook: 0.9.11
  - cospend: 1.4.6
  - dashboard: 7.3.0
  - dav: 1.21.0
  - deck: 1.6.1
  - federatedfilesharing: 1.13.0
  - federation: 1.13.0
  - files: 1.18.0
  - files_pdfviewer: 2.4.0
  - files_rightclick: 1.2.0
  - files_sharing: 1.15.0
  - files_trashbin: 1.13.0
  - files_versions: 1.16.0
  - files_videoplayer: 1.12.0
  - firstrunwizard: 2.12.0
  - groupfolders: 11.1.2
  - impersonate: 1.10.0
  - logreader: 2.8.0
  - lookup_server_connector: 1.11.0
  - maps: 0.1.10
  - nextcloud_announcements: 1.12.0
  - notes: 4.3.1
  - notifications: 2.11.1
  - oauth2: 1.11.0
  - password_policy: 1.13.0
  - photos: 1.5.0
  - privacy: 1.7.0
  - provisioning_api: 1.13.0
  - recommendations: 1.2.0
  - serverinfo: 1.13.0
  - settings: 1.5.0
  - sharebymail: 1.13.0
  - support: 1.6.0
  - survey_client: 1.11.0
  - suspicious_login: 4.2.0-alpha.1
  - systemtags: 1.13.0
  - talk_matterbridge: 1.23.2
  - terms_of_service: 1.9.3
  - text: 3.4.1
  - theming: 1.14.0
  - twofactor_backupcodes: 1.12.0
  - updatenotification: 1.13.0
  - user_status: 1.3.1
  - viewer: 1.7.0
  - workflowengine: 2.5.0
Disabled:
  - accessibility: 1.8.0
  - activity: 2.15.0
  - apporder: 0.13.0
  - audioplayer: 3.2.2
  - bruteforcesettings: 2.2.0
  - encryption
  - files_external
  - onlyoffice: 7.2.1
  - passman: 2.3.5
  - polls: 3.3.0
  - quicknotes: 0.7.2
  - quota_warning: 1.11.0
  - radio: 1.0.3
  - rainloop: 7.1.2
  - ransomware_protection: 1.11.0
  - spreed: 12.1.2
  - tasks: 0.14.2
  - user_ldap
  - weather_status: 1.2.0
  - workflow_media_converter: 1.3.3
  - workflow_pdf_converter: 1.8.0

**Comment by oddmean** [comment]
I've updated to `24 beta 2` and got into the same issue.

**Comment by scottalanmiller** [comment]
Same boat here. Exactly same error. Pretty vanilla install, very simple.

**Comment by marcelklehr** [comment]
The fix is in, once beta 3 is out and I have time, I'll make a release with v24 support.

**Comment by scottalanmiller** [comment]
> The fix is in, once beta 3 is out and I have time, I'll make a release with v24 support.

Awesome, thanks!

**Comment by nickvergessen** [comment]
beta3 is out

**Comment by Githopp192** [comment]
upgrade ok - overall check shows green

some PHP errors: (when calling cron.php)

[PHP] Error: Class OCA\Bookmarks\Service\FileCache contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (OCP\ICache::isAvailable) at /var/www/html/nextcloud/apps/bookmarks/lib/Service/FileCache.php#17

at 2022-04-08T16:00:06+02:00

[cron] Error: Error: Call to undefined method OCA\Survey_Client\BackgroundJobs\MonthlyReport::setTimeSensitivity() at <<closure>>

0. <<closure>>
   OCA\Survey_Client\BackgroundJobs\MonthlyReport->__construct()
1. /var/www/html/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php line 108
   ReflectionClass->newInstanceArgs()
2. /var/www/html/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php line 116
   OC\AppFramework\Utility\SimpleContainer->buildClass()
3. /var/www/html/nextcloud/lib/private/AppFramework/Utility/SimpleContainer.php line 133
   OC\AppFramework\Utility\SimpleContainer->resolve()
4. /var/www/html/nextcloud/lib/private/AppFramework/DependencyInjection/DIContainer.php line 465
   OC\AppFramework\Utility\SimpleContainer->query()
5. /var/www/html/nextcloud/lib/private/ServerContainer.php line 145
   OC\AppFramework\DependencyInjection\DIContainer->queryNoFallback()
6. /var/www/html/nextcloud/lib/private/BackgroundJob/JobList.php line 303
   OC\ServerContainer->query()
7. /var/www/html/nextcloud/lib/private/BackgroundJob/JobList.php line 241
   OC\BackgroundJob\JobList->buildJob()
8. /var/www/html/nextcloud/cron.php line 144
   OC\BackgroundJob\JobList->getNext()

at 2022-04-08T15:30:14+02:00


[PHP] Error: Error: file_put_contents(/media/appdata_ocxjkdoykhhr/preview/a/f/3/3/0/3/f/1897/4096-4096-max.png): failed to open stream: No such file or directory at /var/www/html/nextcloud/lib/private/Files/Storage/Local.php#292 at <<closure>>

 0. <<closure>>
    OC\Log\ErrorHandler::onError()
 1. /var/www/html/nextcloud/lib/private/Files/Storage/Local.php line 292
    file_put_contents()
 2. /var/www/html/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php line 258
    OC\Files\Storage\Local->file_put_contents()
 3. /var/www/html/nextcloud/lib/private/Files/View.php line 1170
    OC\Files\Storage\Wrapper\Wrapper->file_put_contents()
 4. /var/www/html/nextcloud/lib/private/Files/View.php line 706
    OC\Files\View->basicOperation()
 5. /var/www/html/nextcloud/lib/private/Files/Node/Folder.php line 189
    OC\Files\View->file_put_contents()
 6. /var/www/html/nextcloud/lib/private/Files/SimpleFS/NewSimpleFile.php line 133
    OC\Files\Node\Folder->newFile()
 7. /var/www/html/nextcloud/lib/private/Preview/Generator.php line 346
    OC\Files\SimpleFS\NewSimpleFile->putContent()
 8. /var/www/html/nextcloud/lib/private/Preview/Generator.php line 158
    OC\Preview\Generator->getMaxPreview()
 9. /var/www/html/nextcloud/lib/private/Preview/Generator.php line 110
    OC\Preview\Generator->generatePreviews()
10. /var/www/html/nextcloud/lib/private/PreviewManager.php line 212
    OC\Preview\Generator->getPreview()
11. /var/www/html/nextcloud/core/Controller/PreviewController.php line 169
    OC\PreviewManager->getPreview()
12. /var/www/html/nextcloud/core/Controller/PreviewController.php line 142
    OC\Core\Controller\PreviewController->fetchPreview()
13. /var/www/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 225
    OC\Core\Controller\PreviewController->getPreviewByFileId()
14. /var/www/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 133
    OC\AppFramework\Http\Dispatcher->executeController()
15. /var/www/html/nextcloud/lib/private/AppFramework/App.php line 172
    OC\AppFramework\Http\Dispatcher->dispatch()
16. /var/www/html/nextcloud/lib/private/Route/Router.php line 298
    OC\AppFramework\App::main()
17. /var/www/html/nextcloud/lib/base.php line 1023
    OC\Route\Router->match()
18. /var/www/html/nextcloud/index.php line 36
    OC::handleRequest()

GET /core/preview?fileId=1897&x=250&y=250
from xxxx by admin at 2022-04-08T15:40:16+02:00

**Comment by nickvergessen** [comment]
Background job of Survey client is fixed here: https://github.com/nextcloud/survey_client/pull/128

**Comment by Githopp192** [comment]
thx Joas

**Comment by marcelklehr** [comment]
Fixed by e70ccbe3

**Comment by marcelklehr** [comment]
https://github.com/nextcloud/bookmarks/releases/tag/v10.3.0 is now published with support for nc24
